### ISSUE TITLE
saveas does not work with non-bed output - bug or feature?

### ISSUE THREAD

**Issue by adomingues** [daler/pybedtools#146]
I was doing some `groupby` calculations returning a two column results, but saving them with `.saveas()` returns an error:

> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/usr/local/lib/python2.7/dist-packages/pybedtools/bedtool.py", line 668, in decorated
>     result = method(self, *args, **kwargs)
>   File "/usr/local/lib/python2.7/dist-packages/pybedtools/bedtool.py", line 2729, in saveas
>     fn = self._collapse(self, fn=fn, trackline=trackline)
>   File "/usr/local/lib/python2.7/dist-packages/pybedtools/bedtool.py", line 1097, in _collapse
>     for i in iterable:
>   File "/usr/local/lib/python2.7/dist-packages/pybedtools/bedtool.py", line 924, in __iter__
>     elif self.file_type == 'sam':
>   File "/usr/local/lib/python2.7/dist-packages/pybedtools/bedtool.py", line 857, in file_type
>     self._file_type = IntervalFile(self.fn).file_type
>   File "pybedtools/cbedtools.pyx", line 694, in pybedtools.cbedtools.IntervalFile.file_type (pybedtools/cbedtools.cpp:10001)
>   File "pybedtools/cbedtools.pyx", line 695, in pybedtools.cbedtools.IntervalFile.file_type (pybedtools/cbedtools.cpp:9954)
>   File "pybedtools/cbedtools.pyx", line 529, in pybedtools.cbedtools.create_interval_from_list (pybedtools/cbedtools.cpp:7272)
> IndexError: list index out of range

Saving the result with `output=` works fine. My guess is that `.saveas()` tries to validate the output as the bed file, but `output=` doesn't (sorry couldn't find the source code to see if this is correct). Here is a reproducible example:

```python
# Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
# [GCC 4.8.2] on linux2
# Type "help", "copyright", "credits" or "license" for more information.

import pybedtools
# pybedtools 0.6.9

a = pybedtools.example_bedtool('gdc.gff')
b = pybedtools.example_bedtool('gdc.bed')
c = a.intersect(b, c=True)
print 
d = c.groupby(g=[1, 4, 5], c=10, ops=['sum'])
print d
## works
d.saveas('test1.tmp')

## now 2 col output
print c.groupby(g=[3], c=10, ops=['sum'])
## fails
c.groupby(g=[3], c=10, ops=['sum']).saveas('test2.tmp')
## works
c.groupby(g=[3], c=10, ops=['sum'], output='test2.tmp')
``` 

This might be a feature, but I couldn't find it in the docs.

Cheers

**Comment by daler** [comment]
Thanks for reporting this and the example. I'm actually undecided whether this is a bug or not.

(Below is some thinking aloud in case you have further suggestions.
TL;DR: I will add an `as_text` argument to `saveas()`)

---

This issue is related to #110 and #145, where output from BEDTools programs results in non-BED-format files. Sending that incorrectly-formatted file to other tools expecting BED-like input raises an error, as it should.

One "tool" expecting BED-like input is iterating over a `BedTool` object. Since the iterator is designed to yield `Interval` objects, we get errors when it can't create one due to non-BED format. Hence #145.

For this issue #146, `saveas()` is simply calling the iterator, running into similar problems.

Options around this:
1. leave things as-is
2. give `saveas()` a new argument -- maybe `as_text`? -- that just dumps to file without making an iterator out of things. Then using `.saveas(as_text=True)` would solve your issue here, but not ones like #145.
3. have `BedTool` objects be smarter about what they should yield when iterating. If creating an `Interval` out of the first line works, then yield `Interval` objects from then on; otherwise just yield lines. Or lines split on tabs maybe.


Option 1 is OK (of course I'd add more docs about it, and there should really be an FAQ entry on this topic).

Option 2 seems reasonable. While it wouldn't solve #145, it would make the solution to #145 cleaner. That is, you wouldn't have to treat the `BedTool.fn` differently depending on whether it was streaming or file-based.

Option 3 is not very explicit and is likely to create more confusion since the kind of object yielded is not consistent.

So . . . I'll add a new argument to `saveas()`.

**Comment by adomingues** [comment]
I would also be happy with option 1 and call it a feature :) Like you said, adding another option to `saveas()` won't really fix the other issues, and mine is already fixed with the available options. 

My thinking is that  `.saveas()` is expected to be a `BedTool` object, whereas `output` would be the end of that chain of operations by a BED function. So there is already:
a) in-built **robustness** using `.saveas()` which makes sure that the obj is `BedTool`, and 
b) **flexibility** because one can ignore proper bed format and save whatever the result is with `output=`.

Adding a line explaining the difference between `output=`and `.saveas()` in the documentation should suffice.

**Comment by daler** [comment]
OK, [easy enough](http://daler.github.io/pybedtools/topical-saving.html#working-with-non-interval-output-files)!
